### PR TITLE
Bracket changes

### PR DESCRIPTION
**Pull request by alexandru**
Changes meant for finishing https://github.com/typelevel/cats-effect/pull/113

Please merge your branch with `master` from upstream, otherwise you've got changes not related to `Bracket`.

### CODE REVIEW

File: core/shared/src/main/scala/cats/effect/IO.scala [hunk 1]
```diff
@@ -387,8 +387,142 @@ sealed abstract class IO[+A] extends internals.IOBinaryCompat[A] {
...
+   * task gets cancelled, then `bracket` will still perform the
+   * `release`, but the yielded task will be non-terminating
+   * (equivalent with [[IO.never]]).
+   *
+   * Example:
+   *
+   * {{{
+   *   import java.io._
+   *
+   *   def readFile(file: File): Task[String] = {
```
Comment (LukaJCB) [inline core/shared/src/main/scala/cats/effect/IO.scala]: Should we use `IO` here? I can merge this and change it later if you want :)

File: laws/shared/src/main/scala/cats/effect/laws/ConcurrentLaws.scala [hunk 2]
```diff
@@ -25,25 +25,28 @@ import scala.Predef.{identity => id}
...
-    } yield result
+      // Forked execution, allowing us to cancel it later
+      fiber <- F.start(bracketed)
+      // Waits for the `use` action to execute
+      waitStart <- startLatch.await[F]
+      // Triggers cancelation
+      _ <- fiber.cancel
+      // Observes cancelation via bracket's `release`
+      waitExit <- exitLatch.await[F]
+    } yield f(waitStart, waitExit)
```
Comment (LukaJCB) [inline laws/shared/src/main/scala/cats/effect/laws/ConcurrentLaws.scala:47]: Ah, excellent, thanks a lot! :+1:

**Review by LukaJCB — APPROVED** [review]
Thank you so much for your help, I got frustrated formulating that cancellation law, this is really great! Again thank you! :)

File: core/shared/src/main/scala/cats/effect/IO.scala (hunk 1 above)
Comment (alexandru) [reply core/shared/src/main/scala/cats/effect/IO.scala]: Yes, I copy/pasted the ScalaDoc from Monix's Task. Will make changes.

Comment (alexandru) [reply core/shared/src/main/scala/cats/effect/IO.scala]: OK, done.

**Review by LukaJCB — APPROVED** [review]

**Pull request merged**